vimode/src/cmds/motion.c

```diff
@@ -97,18 +108,20 @@ void cmd_goto_up_nonempty(CmdContext *c, CmdParams *p)
 
 static void goto_down(CmdParams *p, gint num)
 {
-	gint one_above, pos;
-	gint last_line = p->line_num - 1;
+	gint previous = -1;
+	gint new_line;
 
-	if (p->line == last_line)
+	if (p->line >= p->line_num - 1)
 		return;
 
-	/* see cmd_goto_up() for explanation */
-	one_above = p->line + num - 1;
-	one_above = one_above < last_line ? one_above : last_line - 1;
-	pos = SSM(p->sci, SCI_GETLINEENDPOSITION, one_above, 0);
-	SET_POS_NOX(p->sci, pos, FALSE);
-	SSM(p->sci, SCI_LINEDOWN, 0, 0);
+	new_line = doc_line_from_visible_delta(p, p->line, num, &previous);
+
+	if (previous > -1) {
+		guint pos = SSM(p->sci, SCI_GETLINEENDPOSITION, previous, 0);
+		SET_POS_NOX(p->sci, pos, FALSE);
+	}
+
+	if (new_line > p->line) SSM(p->sci, SCI_LINEDOWN, 0, 0);
```

Comment:
Is all this code necessary? Maybe I'm missing something but I did just this: 

https://github.com/geany/geany-plugins/pull/1338/commits/fa7025ba9d58fb4680fb47e13bd5c05c6f1f1059#diff-15a3a15a958bc6f85e0f1fae419e23c60bbee47bf617ef133f7539fc1f29b277R128

`doc_line_from_visible_delta()` only returns a line from the valid line range and when you are on the first line, this is already the "line above" for which you then perform `SCI_LINEDOWN` so it should be OK on this side. On the other end of the document when you are on the last line, `SCI_LINEDOWN` just won't do anything so there's no need for some special handling there.

Reply:
It's just a detail but this code is useful to have the same behavior as VIM.
With VIM, when the cursor is on the last line but not the last character and you press the down arrow, the cursor does not move.
Without this code, the cursor will move to the end of the line.

Reply:
Alright, yeah, the previous code handled right that, I just forgot why exactly it was there. I've fixed it in my PR.